File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/vehicle/EngineForceCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.vehicle;

import com.github.aadvorak.artilleryonline.battle.calculations.*;
import com.github.aadvorak.artilleryonline.battle.common.Force;
import com.github.aadvorak.artilleryonline.battle.common.JetType;
import com.github.aadvorak.artilleryonline.battle.common.MovingDirection;
import com.github.aadvorak.artilleryonline.battle.config.VehicleConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;
import com.github.aadvorak.artilleryonline.battle.precalc.VehiclePreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.VehicleSpecs;
import com.github.aadvorak.artilleryonline.battle.state.VehicleState;

import java.util.ArrayList;
import java.util.List;

public class EngineForceCalculator implements ForceCalculator<
        VehicleSpecs,
        VehiclePreCalc,
        VehicleConfig,
        VehicleState,
        VehicleModel,
        VehicleCalculations> {

    private static final String FORCE_DESCRIPTION = "Engine";

    @Override
    public List<BodyForce> calculate(VehicleCalculations calculations, BattleModel battleModel) {
        var forces = new ArrayList<BodyForce>();
        var contactsNumber = calculations.getAllGroundContacts().size();
        var leftWheelForce = calculateForWheel(calculations.getLeftWheel(), contactsNumber);
        var rightWheelForce = calculateForWheel(calculations.getRightWheel(), contactsNumber);
        if (leftWheelForce != null) {
            forces.add(leftWheelForce);
        }
        if (rightWheelForce != null) {
            forces.add(rightWheelForce);
        }
        return forces;
    }

    private BodyForce calculateForWheel(WheelCalculations calculations, int contactsNumber) {
        var vehicleModel = calculations.getModel();
        var jetState = vehicleModel.getState().getJetState();
        var jetSpecs = vehicleModel.getConfig().getJet();
        var direction = vehicleModel.getState().getMovingDirection();
        var contact = calculations.getGroundContact();
        if (direction == null || vehicleModel.getState().getTrackState().isBroken()
                || jetSpecs != null && jetState != null && jetState.isActive() && jetSpecs.getType().equals(JetType.VERTICAL)
                || contact == null) {
            return null;
        }
        var wheelRadius = vehicleModel.getSpecs().getWheelRadius();
        var forceMagnitude = calculations.getMass() * vehicleModel.getSpecs().getAcceleration()
                * Math.cos(contact.angle()) / contactsNumber;
        var depthAngle = contact.depth() * Math.PI / (4 * wheelRadius);
        var force = new Force();
        if (MovingDirection.RIGHT.equals(direction)) {
            force
                    .setX(forceMagnitude * Math.cos(contact.angle() + depthAngle))
                    .setY(forceMagnitude * Math.sin(contact.angle() + depthAngle));

        }
        if (MovingDirection.LEFT.equals(direction)) {
            force
                    .setX( - forceMagnitude * Math.cos(contact.angle() - depthAngle))
                    .setY( - forceMagnitude * Math.sin(contact.angle() - depthAngle));
        }
        return BodyForce.of(force, calculations.getPosition(),
                calculations.getModel().getState().getPosition().getCenter(), FORCE_DESCRIPTION);
    }
}

File: frontend/playground/battle/calculator/vehicle/engine-force-calculator.ts
import type {ForceCalculator} from "~/playground/battle/calculator/force-calculator";
import {type VehicleCalculations, type WheelCalculations} from "~/playground/data/calculations";
import type {BattleModel, VehicleModel} from "~/playground/data/model";
import {BodyForce} from "~/playground/battle/calculator/body-force";
import {JetType, MovingDirection, zeroVector} from "~/playground/data/common";

export class EngineForceCalculator implements ForceCalculator<VehicleCalculations> {
  private static readonly FORCE_DESCRIPTION = 'Engine'

  calculate(calculations: VehicleCalculations, battleModel: BattleModel): BodyForce[] {
    const forces: BodyForce[] = []
    const leftWheelForce = this.calculateForWheel(calculations.leftWheel, calculations.model)
    const rightWheelForce = this.calculateForWheel(calculations.rightWheel, calculations.model)

    if (leftWheelForce !== null) {
      forces.push(leftWheelForce)
    }

    if (rightWheelForce !== null) {
      forces.push(rightWheelForce)
    }

    return forces
  }

  private calculateForWheel(calculations: WheelCalculations, vehicleModel: VehicleModel): BodyForce | null {
    const jetState = vehicleModel.state.jetState
    const jetSpecs = vehicleModel.config.jet
    const direction = vehicleModel.state.movingDirection
    const contact = calculations.groundContact
    if (
        direction === null ||
        vehicleModel.state.trackState.broken ||
        (jetSpecs !== null && jetState !== null && jetState.active && jetSpecs.type === JetType.VERTICAL) ||
        !contact
    ) {
      return null
    }

    const wheelRadius = vehicleModel.specs.wheelRadius
    const forceMagnitude = vehicleModel.preCalc.mass * vehicleModel.specs.acceleration
        * Math.cos(contact.angle) / 2
    const depthAngle = (contact.depth * Math.PI) / (4 * wheelRadius)
    const force = zeroVector()

    if (direction === MovingDirection.RIGHT) {
      force.x = forceMagnitude * Math.cos(contact.angle + depthAngle)
      force.y = forceMagnitude * Math.sin(contact.angle + depthAngle)
    }

    if (direction === MovingDirection.LEFT) {
      force.x = -forceMagnitude * Math.cos(contact.angle - depthAngle)
      force.y = -forceMagnitude * Math.sin(contact.angle - depthAngle)
    }

    return BodyForce.of(
        force,
        contact.position,
        vehicleModel.state.position,
        EngineForceCalculator.FORCE_DESCRIPTION
    )
  }
}
